Run smoke tests on running node

ffakenz opened the pull request:
<!-- Describe your change here -->
## Why
We want to host a hydra-explorer on the same runner used to perform our smoke-tests.

Currently the smoke-tests clean-up the cardano-node/db dir when the mithril flag is being used.

For that reason we need to remove this logic from the CI so that running the smoke-tests does not affect the new hosted explorer.

## What
This PR attempts to adress these concerns by allowing the smoke-tests to:
+ run on an existing running cardano-node instance
+ or to spin-up a new cardano-node instance if there is none.

---

<!-- Consider each and tick it off one way or the other -->
* [X] CHANGELOG updated or not needed
* [X] Documentation updated or not needed
* [X] Haddocks updated or not needed
* [X] No new TODOs introduced or explained herafter

ch1bo commented on hydra-cluster/exe/hydra-cluster/Main.hs:
```diff
@@ -24,13 +26,21 @@ run options =
+              publishOrReuseHydraScripts tracer node
+                >>= singlePartyHeadFullLifeCycle tracer workDir node
+        Nothing -> do
+          removeDirectoryRecursive $ workDir </> "db"
```
Why is this removing of `db` needed now?

ffakenz replied:
because it was removed from the smoke test workflow.

I'll move the moving this operation to above as requested [here](https://github.com/input-output-hk/hydra/pull/1285#discussion_r1480038804)

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -135,21 +147,43 @@ withCardanoNodeDevnet ::
+    Just runningNode -> action runningNode
+
+-- | Run a cardano-node as normal network participant on a known network,
+-- but only if there is no instance already running.
```
Should: fix this haddock

It is always trying to run a cardano-node on known network.

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -135,21 +147,43 @@ withCardanoNodeDevnet ::
+    Nothing -> do
+      args <- setupCardanoDevnet stateDirectory
+      withCardanoNode tracer stateDirectory args action
+    Just runningNode -> action runningNode
```
Must: not need to change this

Why is this change needed?

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -125,6 +125,18 @@ getCardanoNodeVersion :: IO String
+findRunningCardanoNode stateDirectory = do
+  let args'@CardanoNodeArgs{nodeSocket} = defaultCardanoNodeArgs
+  let nodeSocketPath = File nodeSocket
+  maybeSocketPath <- doesFileExist $ unFile nodeSocketPath
```
Must: check also if the node is on the right network, fail otherwise.

There _can_ exist `node.socket` files even if the node does not run. Hence, we can use `node.socket` as a first indication, but we **must** try to communicate with the node to deem it present.

Also (and maybe more important) the node running must be serving the right network id. So we querying the tip or some other basic query to verify the expected `NetworkId` is needed here.

ch1bo commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -507,115 +508,129 @@ spec = around (showLogsOnFailure "EndToEndSpec") $ do
+          maybeRunningNode <- findRunningCardanoNode tmpDir
+          case maybeRunningNode of
+            Just _ -> fail "can't run forking e2e specs on running cardano-node"
+            Nothing -> do
```
Must: not need this check

There is no expectation that these tests would run against an already running cardano node and the `tmpDir` is also randomly selected. Hence, these should just work as before!

Also applies to all changes in this file.

ch1bo commented on hydra-cluster/test/Test/CardanoNodeSpec.hs:
```diff
@@ -41,13 +41,13 @@ spec = do
       showLogsOnFailure "CardanoNodeSpec" $ \tr ->
         withTempDir "hydra-cluster" $ \tmp ->
-          withCardanoNodeOnKnownNetwork tr tmp Mainnet $
+          withNewCardanoNodeOnKnownNetwork tr tmp Mainnet $
```
Could: have built in that check into the existing `withCardanoNodeOnKnownNetwork` and not need to change any names or interfaces at all!

ch1bo commented on hydra-cluster/exe/hydra-cluster/Main.hs:
```diff
@@ -24,13 +26,21 @@ run options =
+            Just node -> do
+              waitForFullySynchronized fromCardanoNode node
+              publishOrReuseHydraScripts tracer node
+                >>= singlePartyHeadFullLifeCycle tracer workDir node
```
Should: avoid duplication of these three lines

ch1bo commented on hydra-cluster/exe/hydra-cluster/Main.hs:
```diff
@@ -24,13 +26,21 @@ run options =
+          case maybeRunningNode of
+            Nothing -> do
+              when (useMithril == UseMithril) $
+                downloadLatestSnapshotTo (contramap FromMithril tracer) network workDir
```
Must: delete the `db/` directory here if you move it out of the workflow

In principle a good idea to not do it in the workflow, but it's nowhere deleted now so this will never work (mithril fails if the `db` dir already exists)

v0d1ch commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -125,6 +125,18 @@ getCardanoNodeVersion :: IO String
   readProcess "cardano-node" ["--version"] ""
 
+-- | Tries to reach for an existing cardano-node running on default state directory.
+findRunningCardanoNode :: FilePath -> IO (Maybe RunningNode)
```
The node.socket file stays in place even if cardano-node is not running. This function is used a lot what would happen if the node is actually down but the file is in place?

ffakenz replied:
Good catch!

And it seems it has been also [pointed out](https://github.com/input-output-hk/hydra/pull/1285/files#r1480024325) by @ch1bo that this solution is not enough.

I'll be tackling this issue as part of the above request, is that ok for you?
would it solve your concern?

ch1bo reviewed (changes requested):
Existence of node.socket is not enough and modification of `withCardanoNodeDevnet` must not be needed.

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -125,6 +125,28 @@ getCardanoNodeVersion :: IO String
+          Mainnet -> Api.Mainnet
+          Preproduction -> Api.Testnet (Api.NetworkMagic 1)
+          Preview -> Api.Testnet (Api.NetworkMagic 2)
+          Sanchonet -> Api.Testnet (Api.NetworkMagic 4)
```
Could: move this next to the `KnownNetwork` data type.

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -125,6 +125,28 @@ getCardanoNodeVersion :: IO String
+  let nodeSocketPath = File nodeSocket
+  maybeSocketPath <- doesFileExist $ unFile nodeSocketPath
+  if maybeSocketPath
+    then do
```
Could: not do the `doesFileExist` at all.

The `try` below will still yield a `Nothing` if the file does not exist. After all, the notion of file existing here is worthless because it can also not exist when you try access it below.

This is "ask for forgiving, instead of allowance".

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -125,6 +125,28 @@ getCardanoNodeVersion :: IO String
+        >>= \case
+          Right _ -> pure Nothing
+          Left (_ :: SomeException) -> do
+            n@RunningNode{networkId} <- mkRunningNode stateDirectory nodeSocketPath args'
```
Should: not need to path both `nodeSocketPath` and `args'` to `mkRunningNode`

The arguments do include the node socket file name, so this is redundant.

You could call `RunningNode{networkId, nodeSocket} <- mkRunningNode ...` in the very beginning and using that `try (queryTip networkId nodeSocket) >>= \case` here. This will simplify the overall call sequence.

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -125,6 +125,28 @@ getCardanoNodeVersion :: IO String
+          Right _ -> pure Nothing
+          Left (_ :: SomeException) -> do
+            n@RunningNode{networkId} <- mkRunningNode stateDirectory nodeSocketPath args'
+            pure $ if knownNetworkId == networkId then Just n else Nothing
```
Sould: use `guard`

ffakenz replied:
unfortunately `guard` is not the same.

how would you use it in this situation?

ch1bo replied:
Ah.. it's `IO (Maybe RunningNode)`. So it would be `pure $ guard (knownNetworkId == networkId) >> pure n`

But this is not much clearer indeed.

ch1bo commented on hydra-cluster/exe/hydra-cluster/Main.hs:
```diff
@@ -24,20 +26,30 @@ run options =
+                  runSmokeTest tracer workDir fromCardanoNode
+            )
+            (runSmokeTest tracer workDir fromCardanoNode)
+            maybeRunningNode
```
```suggestion
          withRunningCardanoNode workDir network $ \node -> do
            waitForFullySynchronized fromCardanoNode node
            publishOrReuseHydraScripts tracer node
              >>= singlePartyHeadFullLifeCycle tracer workDir node
```

and in the `where`:

```
withRunningCardanoNode workDir network action =
  findRunningCardanoNode workDir network >>= \case
    Just node ->
      action node
    Nothing -> do
      when (useMithril == UseMithril) $ do
        removeDirectoryRecursive $ workDir </> "db"
        downloadLatestSnapshotTo (contramap FromMithril tracer) network workDir
      withCardanoNodeOnKnownNetwork fromCardanoNode workDir network action
```

ch1bo commented on hydra-cluster/exe/hydra-cluster/Main.hs:
```diff
@@ -24,20 +26,30 @@ run options =
  where
   Options{knownNetwork, stateDirectory, publishHydraScripts, useMithril} = options
 
+  runSmokeTest tracer workDir fromCardanoNode node = do
```
Should: avoid passing in two tracers

This is a smell of this not being a good function to factor out. Instead the finding or starting a node would be better to factor out. See my suggestion above for a better split.

ch1bo reviewed (approved):
Works.. but please consider my "Should" comments.

v0d1ch commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -259,6 +270,32 @@ forkIntoConwayInEpoch stateDirectory args n = do
 
+mkRunningNode :: FilePath -> CardanoNodeArgs -> IO RunningNode
+mkRunningNode stateDirectory args@CardanoNodeArgs{nodeSocket} = do
+  shelleyGenesis :: Aeson.Value <- readShelleyGenesisJSON $ stateDirectory </> nodeShelleyGenesisFile args
```
Perhaps not needed type annotation?

ffakenz replied:
unfortunately i cant' :(

v0d1ch commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -125,6 +125,17 @@ getCardanoNodeVersion :: IO String
+  try (queryTip knownNetworkId nodeSocket)
+    >>= \case
+      Right _ -> pure Nothing
+      Left (_ :: SomeException) -> do pure (Just n)
```
Perhaps not needed type annotation?

ffakenz replied:
unfortunately i cant' :(

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -125,6 +125,17 @@ getCardanoNodeVersion :: IO String
+  try (queryTip knownNetworkId nodeSocket)
+    >>= \case
+      Right _ -> pure Nothing
+      Left (_ :: SomeException) -> do pure (Just n)
```
Must: flip these two things. 

The `Right` case is that we successfully can use this node (with this network id). So in that case we want to return `Just n`, where i n the `Left` we have a problem to talk to this node and we should return `Nothing`

ffakenz replied:
my bad

ch1bo commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -125,6 +125,17 @@ getCardanoNodeVersion :: IO String
+findRunningCardanoNode stateDirectory knownNetwork = do
+  let knownNetworkId = toNetworId knownNetwork
+  n@RunningNode{networkId, nodeSocket} <- mkRunningNode stateDirectory defaultCardanoNodeArgs
+  guard (knownNetworkId == networkId)
```
Must: not throw here, otherwise you would need to catch an exception around `findRunningCardanoNode`

`guard` in `IO` throws an exception if it's `False`, see https://hackage.haskell.org/package/base-4.19.0.0/docs/src/GHC.Base.html#line-1741

You _can_ do this, but then the signature `IO (Maybe RunningNode)` does not make sense and it should rather be `IO RunningNode`. Then you need to do a `try` or `catch` at the call site (and you can skip the `try` below)

ch1bo reviewed (dismissed):
This is broken

ch1bo commented:
Fixed this and now ready for re-review.

locallycompact commented:
Commit history should not include reversions for code that occurred in the same PR. We should consider forensic quality in mainline a lot more.

ch1bo commented:
Fixed this failing run by manually removing directories on our hosted mac runner: https://github.com/input-output-hk/hydra/actions/runs/7890406705/job/21532504412?pr=1285

```
cd ~/github/_work/hydra/hydra
chmod -R u+w haddocks
rm -rf haddocks
```

ch1bo commented:
> Commit history should not include reversions for code that occurred in the same PR. We should consider forensic quality in mainline a lot more.

@locallycompact _should_ I do this or _must_ I do this?

I'm fine doing it, but didn't so far to not throw away all the work @ffakenz did.

locallycompact commented:
> > Commit history should not include reversions for code that occurred in the same PR. We should consider forensic quality in mainline a lot more.
> 
> @locallycompact _should_ I do this or _must_ I do this?
> 
> I'm fine doing it, but didn't so far to not throw away all the work @ffakenz did.

I say must. Squash and add co-authors.

ghost commented:
> Fixed this failing run by manually removing directories on our hosted mac runner:

Do you know what caused this problem?

locallycompact commented:
> > Fixed this failing run by manually removing directories on our hosted mac runner:
> 
> Do you know what caused this problem?

It could have been my adding self-hosted linux jobs in another branch and that the Mac picked up one of those linux jobs.

ch1bo committed:
Use already running cardano-node in hydra-cluster

At first we tried to read the genesis shelly json files, but we don't
know the location of shelley genesis keys and should also not rely on
knowing that when using an existing cardano-node.

By fetching the genesis parameters we have a chance to compute the right
block time (assuming the node is not in Byron era) and provide a
RunningNode handle without knowing the genesis json paths.

Co-authored-by: Franco Testagrossa <franco.testagrossa@iohk.io>

ch1bo commented:
@locallycompact Done. Please review.

locallycompact commented on hydra-cluster/exe/hydra-cluster/Main.hs:
```diff
@@ -23,21 +25,29 @@ run options =
  where
   Options{knownNetwork, stateDirectory, publishHydraScripts, useMithril} = options
 
+  withRunningCardanoNode tracer workDir network action =
```
Could: have a local signature here

ch1bo replied:
It's a direct wrapper of `withRunningCardanoNodeOnKnownNewtwork` and I don't see much value of it.
